### ISSUE TITLE
Facing issue in installation of test ttsiod-renderer in aarch64 platform

### ISSUE THREAD

**Issue by ossdev07** [phoronix-test-suite/phoronix-test-suite#323]
Facing below issue in aarch64 platform:
phoronix-test-suite install ttsiod-renderer

Phoronix Test Suite v8.6.0m0

    To Install:    pts/ttsiod-renderer-1.7.0

    Determining File Requirements ............................................................................................................
    Searching Download Caches ................................................................................................................

    1 Test To Install
        37MB Of Disk Space Is Needed

    pts/ttsiod-renderer-1.7.0:
        Test Installation 1 of 1
        1 File Needed [4.1 MB / 1 Minute]
        File Found: renderer-2.3b.zip                                                                                                 [4.10MB]
        Installation Size: 36.2 MB
        Installing Test @ 16:17:57
            The installer exited with a non-zero exit status.
            ERROR: lto-wrapper: fatal error: /usr/bin/g++ returned 1 exit status
            LOG: /var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/install-failed.log
Checked install-failed.log file test is failing due to below issue:
**CXXLD    renderer
ScanConverter.h: In member function 'InnerLoop':
ScanConverter.h:117:5: internal compiler error: Segmentation fault
     }
     ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
lto-wrapper: fatal error: g++ returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status**

Please guide me on solving this issue. Thanks in advance.

**Comment by michaellarabel** [comment]
What issue are you facing? What is the output of `phoronix-test-suite debug-install ttsiod-renderer`?

**Comment by ossdev07** [comment]
@michaellarabel Output of debug-install command is similar to that of install-failed.log file. Final result is:
CXXLD    renderer
ScanConverter.h: In member function 'InnerLoop':
ScanConverter.h:117:5: internal compiler error: Segmentation fault
     }
     ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
lto-wrapper: fatal error: /usr/bin/g++ returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
Makefile:384: recipe for target 'renderer' failed
make[2]: *** [renderer] Error 1
make[2]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/src'
Makefile:331: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/src'
Makefile:493: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1

The installer exited with a non-zero exit status.

ERROR: lto-wrapper: fatal error: /usr/bin/g++ returned 1 exit status

LOG: /var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/install-failed.log

**Comment by michaellarabel** [comment]
" internal compiler error: Segmentation fault" looks like the test is hitting a bug within your compiler.

**Comment by ossdev07** [comment]
@michaellarabel How to find bug which it is hitting and how to resolve it to install this test?

**Comment by michaellarabel** [comment]
As shown in that log file:

> 
> ScanConverter.h:117:5: internal compiler error: Segmentation fault
> }
> ^
> Please submit a full bug report,
> with preprocessed source if appropriate.
> See file:///usr/share/doc/gcc-7/README.Bugs for instructions.

This is a bug with your compiler (GCC) outside of the scope of the Phoronix Test Suite. You would need to open a bug with the GCC developers and/or the TTSIOD developer as the issue is happening within your compiler and not something within PTS side.

**Comment by ossdev07** [comment]
@michaellarabel Thanks for the help. Will raise a bug for the same.

**Comment by ttsiodras** [comment]
@ossdev07 I am the author of the renderer. Have you opened a ticket in GCC regarding this? 

If not, can you point me to the exact version of gcc that you used?  I will then try to reproduce the issue, and open a ticket in GCC myself (the bug is definitely in GCC, not in phoronix or in my renderer).

Thanks in advance!
Thanassis.

**Comment by ossdev07** [comment]
@ttsiodras No, I haven't opened a ticket in gcc regarding this as i was not sure whether this was coming in renderer or gcc. Gcc version which i have used is **gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/aarch64-linux-gnu/7/lto-wrapper
Target: aarch64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro 7.3.0-27ubuntu1~18.04' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=aarch64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libquadmath --disable-libquadmath-support --enable-plugin --enable-default-pie --with-system-zlib --enable-multiarch --enable-fix-cortex-a53-843419 --disable-werror --enable-checking=release --build=aarch64-linux-gnu --host=aarch64-linux-gnu --target=aarch64-linux-gnu
Thread model: posix
gcc version 7.3.0 (Ubuntu/Linaro 7.3.0-27ubuntu1~18.04)
**. 
Thanks.

**Comment by ttsiodras** [comment]
@ossdev07 @michaellarabel Since I don't have access to a 64bit ARM machine, I downloaded a QEMU-based Aarch64 image from [here](https://blahcat.github.io/2017/06/25/qemu-images-to-play-with/). It contained Ubuntu - I SSH-ed inside it and after updating /etc/apt/sources.list to add bionic in the supported versions, proceeded to install the exact same GCC version you reported:

    $ ssh -p 22055 root@localhost
    Last login: Sat Jan 12 12:30:10 2019 from 10.0.2.2
    root@ubuntu1604-aarch64:~# uname -a
    Linux ubuntu1604-aarch64 4.4.0-81-generic #104-Ubuntu SMP Wed Jun 14 08:22:02 UTC 2017 aarch64 aarch64 aarch64 GNU/Linux

    root@ubuntu1604-aarch64:~# gcc -v
    Using built-in specs.
    COLLECT_GCC=gcc
    COLLECT_LTO_WRAPPER=/usr/lib/gcc/aarch64-linux-gnu/7/lto-wrapper
    Target: aarch64-linux-gnu
    Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro 7.3.0-27ubuntu1~18.04' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=aarch64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libquadmath --disable-libquadmath-support --enable-plugin --enable-default-pie --with-system-zlib --enable-multiarch --enable-fix-cortex-a53-843419 --disable-werror --enable-checking=release --build=aarch64-linux-gnu --host=aarch64-linux-gnu --target=aarch64-linux-gnu
    Thread model: posix
    gcc version 7.3.0 (Ubuntu/Linaro 7.3.0-27ubuntu1~18.04) 

And I then proceeded to build my renderer - which built and linked fine...

    $ ./configure
    ...
    $ make
    ...
    Loader.cc:237:22: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
                 fread(&z,1,4,fp); if (feof(fp)) { fclose(fp); THROW("Malformed 3D file"); }
                 ~~~~~^~~~~~~~~~~
    Loader.cc:238:22: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
                 fread(&x,1,4,fp); if (feof(fp)) { fclose(fp); THROW("Malformed 3D file"); }
                 ~~~~~^~~~~~~~~~~
      CXX      renderer-Raytracer.o
      CXXLD    renderer
      CXX      showShadowMap-showShadowMap.o
      CXX      showShadowMap-Keyboard.o
      CXXLD    showShadowMap
    make[2]: Leaving directory '/root/renderer/src'
    make[1]: Leaving directory '/root/renderer/src'
    make[1]: Entering directory '/root/renderer'
    make[1]: Nothing to be done for 'all-am'.
    make[1]: Leaving directory '/root/renderer'

I am afraid I can't reproduce the crash you reported - even though I used the exact same GCC version under the closest SW (OS/GCC version) and "HW" (QEMU/Aarch64) platform I could get access to.
 
The only other potential I can think of is that of a HW issue/instability - e.g. insufficient power supply to the board, thermal issues, etc.

P.S. Note also that last year I had for a brief time SSH access to a Rock64 machine of a friend, and compiled, built and run the renderer fine there.

**Comment by ossdev07** [comment]
@ttsiodras Machine configuration:
**uname -a
Linux e7fac9d4ba1c 4.10.0-38-generic #42~16.04.1-Ubuntu SMP Tue Oct 10 16:33:57 UTC 2017 aarch64 aarch64 aarch64 GNU/Linux**
**gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/aarch64-linux-gnu/7/lto-wrapper
Target: aarch64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro 7.3.0-27ubuntu1 18.04' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=aarch64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libquadmath --disable-libquadmath-support --enable-plugin --enable-default-pie --with-system-zlib --enable-multiarch --enable-fix-cortex-a53-843419 --disable-werror --enable-checking=release --build=aarch64-linux-gnu --host=aarch64-linux-gnu --target=aarch64-linux-gnu
Thread model: posix
gcc version 7.3.0 (Ubuntu/Linaro 7.3.0-27ubuntu1~18.04)
**

I am still getting below error while running make command:
root@e7fac9d4ba1c:/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b# make
Making all in lib3ds-1.3.0
make[1]: Entering directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0'
make  all-recursive
make[2]: Entering directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0'
Making all in lib3ds
make[3]: Entering directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0/lib3ds'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0/lib3ds'
Making all in tools
make[3]: Entering directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0/tools'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0/tools'
make[3]: Entering directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0'
make[3]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0'
make[2]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0'
make[1]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/lib3ds-1.3.0'
Making all in src
make[1]: Entering directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/src'
make  all-am
make[2]: Entering directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/src'
  CXXLD    renderer
ScanConverter.h: In member function 'InnerLoop':
ScanConverter.h:117:5: internal compiler error: Segmentation fault
     }
     ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
lto-wrapper: fatal error: g++ returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
Makefile:384: recipe for target 'renderer' failed
make[2]: *** [renderer] Error 1
make[2]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/src'
Makefile:331: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/var/lib/phoronix-test-suite/installed-tests/pts/ttsiod-renderer-1.7.0/renderer-2.3b/src'
Makefile:493: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1

Do we need to install any dependency? Please suggest on solving this. Thanks in advance.

**Comment by ttsiodras** [comment]
@ossdev07 Thank you for repeating the output - I checked it again, it is identical to the version I used for my test (which didn't reproduce the crash).  Note that both Michael and myself have stated clearly in our previous messages that it has nothing to do with dependencies of any kind: simply put, in your machine, invoking GCC at some point during the build process **the compiler itself crashes** *(and even goes on to suggest debugging this crash via generation of preprocessed output of the source code, and submission to the GCC developers)*.  

[I have done this in the past](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=49223) - which led to a fix in GCC. But since this time I could not reproduce the crash with an identical (SW-wise) setup (i.e. the emulated Aarch64 running the same GCC),  the only way forward that I can see is if you provide me with remote (SSH) access to this machine, so I can debug this further. My theory is that this crash has something to do with your specific HW setup.

To avoid spamming Michael's ticketing system - this has nothing to do with either Phoronix or my renderer - we can continue this over [e-mail](ttsiodras@gmail.com).
